docs: add gadgets to o1js (#737)
o1-labs/docs2: docs/zkapps/o1js/bitwise-operations.mdx

- A larger `length` parameter adds more constraints.

With `length = 2` (`paddedLength = 16`), the `and()` operation fails for any input that is larger than `2**16`.

Example:

```ts
let a = Field(3);    // ... 000011
let b = Field(5);    // ... 000101
let c = Gadgets.and(a, b, 2);    // ... 000001
c.assertEquals(1);
```

## not()

```ts
not(a: Field, length: number, checked: boolean) => Field
```

The bitwise `not()` gadget on Field elements is similar to the [Bitwise NOT (~)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_NOT) operator in JavaScript. It is a provable method to support bitwise shifting for native Field elements.

A bitwise NOT operation returns 1 in each bit position if the corresponding bit of the operand is 0 and returns 0 if the corresponding bit of the operand is 1.

The operation:

- Fails if the length or the input value is larger than 254 bits.
- Operates only over the number of bits specified by the `length` parameter.
    - The `length` parameter defines how many bits to NOT.
    - A larger `length` parameter adds more constraints.
- Is implemented in two different ways:

>     - If the `checked` parameter is set to `false`, it is implemented as a subtraction of the input from the all one bitmask. If the `checked` parameter is not provided, the default implementation uses `false`.

Reviewer: "from the all one bitmask" seems like an error but I don't know what it should say

Example:

```ts
const x = Provable.witness(Field, () => Field(12345678n));
Gadgets.rangeCheck64(x); // successfully proves 64-bit range

const xLarge = Provable.witness(Field, () => Field(12345678901234567890123456789012345678n));
Gadgets.rangeCheck64(xLarge); // throws an error since input exceeds 64 bits
```

## multiRangeCheck()

```ts
multiRangeCheck([x, y, z])
```

The `multiRangeCheck()` gadget is a building block for non-native arithmetic with BigInt of size up to 264 bits. It is a provable method for efficient 64-bit range checks using lookup tables.

The `multiRangeCheck()` gadget:

- Takes 4 rows and checks `88*3/4 = 66` bits per row, so is slightly more efficient than [64-bit range checks](#rangecheck64).
- Supports BigInts up to 264 bits (`3x88-bit`) that supports foreign field multiplication with moduli up to `2^259`.
- Requires input values less than 88 bits. Throws an error if input values exceed 88 bits.

Example:

```ts
multiRangeCheck(limbs: Field3) {

>     multiRangeCheck(limbs);

Reviewer: ```suggestion
  multiRangeCheck(limbs);
```

---

> title: Bitwise Gadgets

Reviewer 2: ```suggestion
title: Bitwise Operations
```

hide_title: true
description: Bitwise operations in o1js are implemented as gadgets.
keywords:
  - gadgets
  - bitwise operations
  - o1js
---

:::info

Please note that zkApp programmability is not yet available on Mina Mainnet, but zkApps can now be deployed to Berkeley Testnet.

:::

# Bitwise Operations

In o1js, bitwise operations are implemented as [gadgets](/zkapps/o1js/gadgets).

- [and()](#and)
- [not()](#not)
- [xor()](#xor)
- [leftShift()](#leftshift)
- [rightShift()](#rightshift)
- [rotate()](#rotate)
- [rangeCheck64()](#rangecheck64)
- [multiRangeCheck()](#multirangecheck)
- [compactMultiRangeCheck()](#compactmultirangecheck)

See the type declaration for [Gadgets](/zkapps/o1js-reference/modules#gadgets) in the o1js Reference documentation.

## and()

```ts
and(a: Field, b: Field, length: number) => Field
```

The bitwise `and()` gadget on Field elements is equivalent to the [bitwise AND (&)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_AND) operator in JavaScript. 

The `and()` gadget:

> - Compares two bits and returns 1 only if both bits are 1 and returns 0 if both bits are not 1.

Reviewer: ```suggestion
- Compares two bits and returns 1 only if both bits are 1. Returns 0 if either bit is not 1.
```

---
title: Bitwise Gadgets
hide_title: true
description: Bitwise operations in o1js are implemented as gadgets.
keywords:
  - gadgets
  - bitwise operations
  - o1js
---

:::info

Please note that zkApp programmability is not yet available on Mina Mainnet, but zkApps can now be deployed to Berkeley Testnet.

:::

# Bitwise Operations

In o1js, bitwise operations are implemented as [gadgets](/zkapps/o1js/gadgets).

- [and()](#and)
- [not()](#not)
- [xor()](#xor)
- [leftShift()](#leftshift)
- [rightShift()](#rightshift)
- [rotate()](#rotate)
- [rangeCheck64()](#rangecheck64)
- [multiRangeCheck()](#multirangecheck)
- [compactMultiRangeCheck()](#compactmultirangecheck)

See the type declaration for [Gadgets](/zkapps/o1js-reference/modules#gadgets) in the o1js Reference documentation.

## and()

> ```ts

Author: I grabbed this "syntax" from http://localhost:3000/zkapps/o1js-reference/modules#gadgets but it's really not a true command syntax, more of an explainer

Did we want to provide literal syntax here? or omit this for each gadget?

and(a: Field, b: Field, length: number) => Field
```

The bitwise `and()` gadget on Field elements is equivalent to the [bitwise AND (&)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_AND) operator in JavaScript. 

> The `and()` gadget:

Author: @garwalsh  and I talked about how we can "write" gadget names and landed on following how we write about methods, using tick marks and parentheses. Does this Markdown syntax work for the docs?

whatever we do, let's do it consistently throughout

Reviewer 2: The proposed convention is:
- When referring to the gadgets themselves we use "and()" to reflect how they appear in code
- When referring to a boolean operation we use AND

The operation:

- Fails if the length or the input value is larger than 254 bits.
- Operates only over the number of bits specified by the `length` parameter.
    - The `length` parameter defines how many bits to NOT.
    - A larger `length` parameter adds more constraints.
- Is implemented in two different ways:
    - If the `checked` parameter is set to `false`, NOT is implemented as a subtraction of the input from the all one bitmask. If the `checked` parameter is not provided, the default implementation uses `false`.
    - If the `checked` parameter is set to `true`, the [xor()](#xor) gadget is reused with a second argument to be an all one bitmask the same length. This approach needs as many rows as an XOR requires for a single negation. 

For implementation details, see [NOT](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#not) in the Mina book.

Example NOT-ing 4 bits with the unchecked version:

```ts
let a = Field(0b0101);
let b = Gadgets.not(a,4,false);

b.assertEquals(0b1010);
```

Example NOT-ing 4 bits with the checked version using `xor()`:

```ts
let a = Field(0b0101);
let b = Gadgets.not(a,4,true);

b.assertEquals(0b1010);
```

> where:

Author: do we need a similar where statement for the examples, or remove this one?

The `rightShift()` gadget:

- Performs a right shift operation on a Field element.
- Shifts bits to the right and discards the overflowing bits.
- Performs these operations with the big-endian 64-bit representation of the number, where the most significant (64th) bit is on the left end and the least significant bit is on the right end.
- Requires values that are range checked to 64 bits. You can use [rangeCheck64](#rangecheck64).

Example:
```ts
const x = Provable.witness(Field, () => Field(0b001100)); // 12 in binary
const y = Gadgets.rightShift(x, 2); // right shift by 2 bits
y.assertEquals(0b000011); // 3 in binary

const xLarge = Provable.witness(Field, () => Field(12345678901234567890123456789012345678n));
rightShift(xLarge, 32); // throws an error since input exceeds 64 bits
```

## rotate()

```ts
rotate(field: Field, bits: number, direction: 'left' | 'right' = 'left') {
return rotate(field, bits, direction);
},
```

> The `rotate()` gadget operates like the [Right shift (>>)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Right_shift) and [Left shift (<<)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Left_shift) operators in JavaScript. 

Author: syntax is correct, I think we can ignore the weird highlighting here

---
title: Bitwise Operations
hide_title: true
description: Bitwise operations in o1js are implemented as gadgets.
keywords:
  - gadgets
  - bitwise operations
  - o1js
---

:::info

Please note that zkApp programmability is not yet available on Mina Mainnet, but zkApps can now be deployed to Berkeley Testnet.

:::

# Bitwise Operations

> Bitwise operations manipulate individual bits within a binary represenbtation of a number. They can at times resemble boolean operations but apply to a sequence of bits instead of booleans. Bitwise operations are generally available in most programming languages, including TypeScript. o1js provides versions of them that operate on `Field` elements and result in the necessary circuit constraints to generate a zero knowledge proof of the computation. This is especially useful when implementing hashing algorithms such as SHA256.

Reviewer 2: ```suggestion
Bitwise operations manipulate individual bits within a binary representation of a number. They can at times resemble boolean operations but apply to a sequence of bits instead of booleans. Bitwise operations are generally available in most programming languages, including TypeScript. o1js provides versions of them that operate on `Field` elements and result in the necessary circuit constraints to generate a zero knowledge proof of the computation. This is especially useful when implementing hashing algorithms such as SHA256.
```

The input `Field` should not 64 bits in size. You can use [rangeCheck64](#rangecheck64) to ensure this.

For implementation details, see [ROTATION](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#rotation) in the Mina book.

Example:

```ts
const x = Provable.witness(Field, () => Field(0b001100));
const y = Gadgets.rotate64(x, 2, 'left'); // left rotation by 2 bits
const z = Gadgets.rotate64(x, 2, 'right'); // right rotation by 2 bits
y.assertEquals(0b110000);
z.assertEquals(0b000011);

const xLarge = Provable.witness(Field, () => Field(12345678901234567890123456789012345678n));
Gadgets.rotate64(xLarge, 32, "left"); // throws an error since input exceeds 64 bits
```

## addMod32()

```ts
addMod32(a: Field, b: Field) => Field
```

The `addMod32()` helper performs addition that overflows on 32-bit numbers, much like the `int32` type. It returns the result of addition modulo `2^32` in a new `Field` element.

The input `Field`s should not 64 bits in size. You can use [rangeCheck64](#rangecheck64) to ensure this.

Example:

```ts
let a = Field(8n);
let b = Field(1n << 32n);

Gadgets.addMod32(a, b).assertEquals(Field(8n));
```

## divMod32()

```ts

> divMod32(field: Field) => [Field, Field]

Reviewer 3: ```suggestion
divMod32(field: Field) => { remainder: Field, quotient: Field }
```

- [rangeCheck32()](#rangecheck32)
- [rangeCheck64()](#rangecheck64)
- [multiRangeCheck()](#multirangecheck)
- [compactMultiRangeCheck()](#compactmultirangecheck)

## and()

```ts
and(a: Field, b: Field, length: number) => Field
```

The bitwise `and()` gadget is a provable equivalent to the [bitwise AND (&)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_AND) operator in JavaScript. It receives two `Field` elements and compares the corresponding pairs of bits from the binary representation of each. The comparison returns 1 only if both bits are 1 and returns 0 if either bit is not 1. This results in a new binary number which is returned as a `Field` element.

For details about the implementation, see [AND](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#and) in the Mina book.

The `length` parameter:

- Specifies how many bits to compare.
- Adds more constraints for larger numbers.

Example:

```ts
let a = Field(3);    // ... 000011
let b = Field(5);    // ... 000101
let c = Gadgets.and(a, b, 2);    // ... 000001
c.assertEquals(1);
```

## not()

```ts
not(a: Field, length: number, checked: boolean) => Field
```

> The bitwise `not()` gadget is a provable equivalent to the [Bitwise NOT (~)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_NOT) operator in JavaScript. It receives a `Field` element and negates each bit of its binary representation, turning all the `1`s into `0`s and all the `0`s into `1`s. It essentially flips all the bits in a `Field` element. This results in a new binary number which is returned as a `Field` element.

Author: so we want to use the tick marks for 1s and 0s here? Seems like we can use plain text 

(love the r/w @nicc !)

The `length` parameter:

- Specifies how many bits to compare.
- Adds more constraints for larger numbers.

For details about the implementation, see [XOR](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#xor-1) in the Mina book.

Example:

```ts
let a = Field(0b0101);
let b = Field(0b0011);

let c = Gadgets.xor(a, b, 4); // xor-ing 4 bits
c.assertEquals(0b0110);
```

## leftShift32()

```ts
leftShift32(field: Field, bits: number) => Field
```

The `leftShift32()` gadget is a provable equivalent to the [Left shift (<<)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Left_shift) operator in JavaScript. It moves the bits of a binary number to the left by the specified number of `bits`. Any bits that fall off the left side are discarded. `0`s are padded in from the right. It returns a new `Field` element that is range checked to 32 bits.

> The input `Field` should not 32 bits in size. You can use [rangeCheck32](#rangecheck32) to ensure this.

Author: ```suggestion
The input `Field` must not exceed 32 bits in size. You can use [rangeCheck32](#rangecheck32) to ensure this.
```

Example:

```ts
const x = Provable.witness(Field, () => Field(0b001100)); // 12 in binary
const y = Gadgets.leftShift32(x, 2); // left shift by 2 bits
y.assertEquals(0b110000); // 48 in binary
```

## leftShift64()

```ts
leftShift64(field: Field, bits: number) => Field
```

The `leftShift64()` gadget is a provable equivalent to the [Left shift (<<)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Left_shift) operator in JavaScript. It moves the bits of a binary number to the left by the specified number of `bits`. Any bits that fall off the left side are discarded. `0`s are padded in from the right. It returns a new `Field` element that is range checked to 64 bits.

> The input `Field` should not 64 bits in size. You can use [rangeCheck64](#rangecheck64) to ensure this.

Author: ```suggestion
The input `Field` must not exceed 64 bits in size. You can use [rangeCheck64](#rangecheck64) to ensure this.
```

Example:

```ts
const x = Provable.witness(Field, () => Field(0b001100)); // 12 in binary
const y = Gadgets.leftShift64(x, 2); // left shift by 2 bits
y.assertEquals(0b110000); // 48 in binary

const xLarge = Provable.witness(Field, () => Field(12345678901234567890123456789012345678n));
Gadgets.leftShift64(xLarge, 32); // throws an error since input exceeds 64 bits
```

## rightShift64()

```ts
rightShift64(field: Field, bits: number) => Field
```

The `rightShift64()` gadget is a provable equivalent to the [Right shift (>>)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Right_shift) operator in JavaScript. It moves the bits of a binary number to the right by the specified number of `bits`. Any bits that fall off the right side are discarded. `0`s are padded in from the left. It returns a new `Field` element.

> The input `Field` should not 64 bits in size. You can use [rangeCheck64](#rangecheck64) to ensure this.

Author: ```suggestion
The input `Field` must not exceed 64 bits in size. You can use [rangeCheck64](#rangecheck64) to ensure this.
```

same comment as above
"must not" or "cannot" (but not should)

let's add "exceeds"

The input `Field` should not 32 bits in size. You can use [rangeCheck32](#rangecheck32) to ensure this.

For implementation details, see [ROTATION](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#rotation) in the Mina book.

Example:

```ts
const x = Provable.witness(Field, () => Field(0b001100));
const y = Gadgets.rotate32(x, 2, 'left'); // left rotation by 2 bits
const z = Gadgets.rotate32(x, 2, 'right'); // right rotation by 2 bits
y.assertEquals(0b110000);
z.assertEquals(0b000011);

const xLarge = Provable.witness(Field, () => Field(12345678901234567890123456789012345678n));
Gadgets.rotate32(xLarge, 32, "left"); // throws an error since input exceeds 32 bits
```

## rotate64()

```ts
rotate64(field: Field, bits: number, direction: 'left' | 'right' = 'left') {
  return rotate64(field, bits, direction);
},
```

> The `rotate64()` gadget performs provable bit rotation on 32 bit numbers. It is similar to left- and right shift, except the bits that fall off the end will wrap around to reappear on the opposite side instead of being discarded. It accepts a `Field` element, the number of `bits` to rotate, and a `direction` of 'left' or 'right'. The default direction is 'left'.

Author: to improve readability, I think plain text works for the description. If we want to use code literals, then let's use tick marks

The input `Field` should not 64 bits in size. You can use [rangeCheck64](#rangecheck64) to ensure this.

For implementation details, see [ROTATION](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#rotation) in the Mina book.

Example:

```ts
const x = Provable.witness(Field, () => Field(0b001100));
const y = Gadgets.rotate64(x, 2, 'left'); // left rotation by 2 bits
const z = Gadgets.rotate64(x, 2, 'right'); // right rotation by 2 bits
y.assertEquals(0b110000);
z.assertEquals(0b000011);

const xLarge = Provable.witness(Field, () => Field(12345678901234567890123456789012345678n));
Gadgets.rotate64(xLarge, 32, "left"); // throws an error since input exceeds 64 bits
```

## addMod32()

```ts
addMod32(a: Field, b: Field) => Field
```

The `addMod32()` helper performs addition that will overflow on 32 bit numbers, much like the `int32` type. It returns the result of addition modulo `2^32` in a new `Field` element.

> The input `Field`s should not 64 bits in size. You can use [rangeCheck64](#rangecheck64) to ensure this.

Reviewer 3: changing to rangeCheck32 which is correct
```suggestion
The input `Field`s must not exceed 32 bits in size. You can use [rangeCheck32](#rangecheck32) to ensure this.
```

---
title: Bitwise Operations
hide_title: true
description: Bitwise operations in o1js are implemented as gadgets.
keywords:
  - gadgets
  - bitwise operations
  - o1js
---

:::info

Please note that zkApp programmability is not yet available on Mina Mainnet, but zkApps can now be deployed to Berkeley Testnet.

:::

# Bitwise Operations

In o1js, bitwise operations are implemented as [gadgets](/zkapps/o1js/gadgets).

- [and()](#and)
- [not()](#not)
- [xor()](#xor)
- [addMod32()](#addmod32)
- [divMod32()](#divmod32)
- [leftShift32()](#leftshift32)
- [leftShift64()](#leftshift64)
- [rightShift32()](#rightshift32)
- [rightShift64()](#rightshift64)
- [rotate32()](#rotate32)
- [rotate64()](#rotate64)

> - [rangeCheck32()](#rangecheck32)

Reviewer 3: FYI - some of these aren't in the current release as planned, but in this PR: https://github.com/o1-labs/o1js/pull/1259
(but maybe we can merge the PR before release, cc @Trivo25)

Reviewer 4: The PR is ready from my side, except two or three tiny follow ups - I am just waiting on the crypto stamp, unless you think thats not needed

- [rangeCheck64()](#rangecheck64)
- [multiRangeCheck()](#multirangecheck)
- [compactMultiRangeCheck()](#compactmultirangecheck)

## and()

```ts
and(a: Field, b: Field, length: number) => Field
```

> The bitwise `and()` gadget on Field elements is equivalent to the [bitwise AND (&)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_AND) operator in JavaScript. 

Author: Is `and()` a provable method? If yes, let's add

Author: is and() only for native Field elements? (or maybe all of these bitwise operations are for native field elements?)
I feel like we can clarify

Reviewer: They are indeed for Field elements. This can be seen in [the function definition](https://github.com/o1-labs/o1js/blob/4efb9e505ba12170cbc10c372815e9532aafc77f/src/lib/gadgets/bitwise.ts#L153). It accepts Field types as inputs. And yes anything provable has to work on Field types so this applies to all of these operations. We only need to reimplement them because they have to be provable. They are well established and already exist in a non-zk context / not on Field types.

Reviewer 3: * all of these are provable methods, so maybe good to highlight that more broadly in the introduction, instead of on each method. (re what is it proving - it's proving that you computed the bitwise AND of two numbers)
* all of these operate on native field elements (`Field`). I wouldn't necessarily highlight that on each method because `Field` is kind of the "default type" in provable code

> The bitwise `and()` gadget on Field elements is equivalent to the [bitwise AND (&)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_AND) operator in JavaScript. 

Reviewer: Yes I believe it is a provable method, @barriebyron. Here's a suggested edit. Eng. reviewers can confirm but I'm pretty sure.
```suggestion
The bitwise `and()` gadget on Field elements is equivalent to the [bitwise AND (&)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_AND) operator in JavaScript. It is a provable method for bitwise and operations.
```

    - If the `checked` parameter is set to `true`, the [xor()](#xor) gadget is reused with a second argument to be an all-ones bitmask of the same length. This approach is less efficient because it needs as many rows as an XOR requires for a single negation, but comes with the advantage of making sure the input is of a certain length.

For implementation details, see [NOT](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#not) in the Mina book.

Example NOT-ing 4 bits with the unchecked version:

```ts
let a = Field(0b0101);
let b = Gadgets.not(a,4,false);

b.assertEquals(0b1010);
```

Example NOT-ing 4 bits with the checked version using `xor()`:

```ts
let a = Field(0b0101);
let b = Gadgets.not(a,4,true);

b.assertEquals(0b1010);
```

## xor()

```ts
xor(a: Field, b: Field, length: number) => Field
```

The `xor()` gadget is equivalent to the [Bitwise XOR (^)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_XOR) operator in JavaScript. It is a provable method to support bitwise XOR operations for native Field elements. 

The XOR operation:

> - Compares two bits and returns 1 if two bits differ. Returns 0 if the two bits are equal.

Reviewer: ```suggestion
- Compares two bits and returns 1 if they differ. Returns 0 if the two bits are equal.
```

- Verifies that each XOR gate can verify at most 16 bits.
- Adds another XOR gate to the chain if your input elements exceed 16 bits.
- Rounds `length` to the nearest multiple of 16, `paddedLength = ceil(length / 16) * 16`.
- Constrains input values to fit into `paddedLength` bits. 
- Requires that both Field elements fit into `2^paddedLength`.
    For example, `length = 2` (`paddedLength = 16`) fails for any input that is larger than `2**16`.
- Guarantees the output has at most `paddedLength` bits.

> The `length` parameter defines how many bits to compare. When choosing the value of the `length` parameter, be sure to carefully weigh the trade-off between an increased number of constraints and security:

Author: we have a variety of ways we talk about the length parameter. Would it make sense to expand the length parameter discussion for the other gadgets to be more like this one?

Reviewer: We'd need to confirm whether all length parameters result in additional constraints with larger values. Will defer to eng. review on this one.

The `length` parameter defines how many bits to compare. When choosing the value of the `length` parameter, be sure to carefully weigh the trade-off between an increased number of constraints and security:
- A larger `length` parameter adds more constraints.
- A smaller `length` allows the verifier to infer the length of the original input data. For example, it will be smaller than 16 bits if only one XOR gate is used.

For details about the implementation, see [XOR](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#xor-1) in the Mina book.

Example:

```ts
let a = Field(0b0101);
let b = Field(0b0011);

let c = Gadgets.xor(a, b, 4); // xor-ing 4 bits
c.assertEquals(0b0110);
```

## addMod32()

```ts
addMod32(a: Field, b: Field) => Field
```

The `addMod32()` operation adds two Field elements in the range `[0, 2^64]` and returns the result modulo `2^32`. It is a helper method to interact with 3-limb vectors of Fields.

The operation:

> - Is addition modulo `2^32`.

Author: Do we need to describe what this is?

Reviewer: This is actually (perhaps surprisingly) specific. I suspect that people using this will understand it. afaik it's an efficient way to do addition that is overflow safe. All addition in SHA256 works this way. Will defer to eng. review on whether there's an easy way to explain it here cos that would be nice but I don't think we should go into depth on it here.

- Asserts that the result is in the range `[0, 2^32)` using [rangeCheck32()](#rangeCheck32).

> - Uses [divMod32()](#divmod32) internally by adding the two Field elements,  decomposing the result into `remainder` and `quotient`, and returning the `remainder`.

Reviewer 3: too much internal detail IMO

- Requires both inputs to be in the range `[0, 2^64)`. 
  - If the operation is called with non-range-checked inputs, the sum `a + b` can overflow the native field and the gadget can succeed but return an invalid result.

Example:

```ts
let a = Field(8n);
let b = Field(1n << 32n);

Gadgets.addMod32(a, b).assertEquals(Field(8n));
```

## divMod32()

```ts

> divMod32(a: Field, b: Field) => Field

Reviewer: As above, could be good to know when this helper is useful.

y.assertEquals(0b110000); // 48 in binary
```

## leftShift64()

```ts
leftShift64(field: Field, bits: number) => Field
```

The `leftShift64()` gadget is a provable method that supports the bitwise shifting operation that moves the bits of a binary number to the left. Unlike rotation, the bits that fall off at the end are discarded and the vacant positions are filled with zeros.

The operation:

- Performs a left shift operation on a Field element.
- Shifts bits to the left and discards the overflowing bits.
- Performs these operations with the big-endian 64-bit representation of the number, where the most significant (64th) bit is on the left end and the least significant bit is on the right end.
- Requires values that are range checked to 64 bits. You can use [rangeCheck64](#rangecheck64).

Example:

```ts
const x = Provable.witness(Field, () => Field(0b001100)); // 12 in binary
const y = Gadgets.leftShift(x, 2); // left shift by 2 bits
y.assertEquals(0b110000); // 48 in binary

const xLarge = Provable.witness(Field, () => Field(12345678901234567890123456789012345678n));
leftShift(xLarge, 32); // throws an error since input exceeds 64 bits
``

> ## leftShift32()

Reviewer: This is repeated. Looks like copy-pasta

> ## leftShift32()

Reviewer: ```suggestion
## rightShift32()
```

Looks like a whole lot of copy-pasta. I'll provide inline suggestions to correct on the assumption that it's the same but right-leaning.

```ts

> leftShift32(field: Field, bits: number) => Field

Reviewer: ```suggestion
rightShift32(field: Field, bits: number) => Field
```

```

The `leftShift32()` gadget is a provable method that supports the bitwise shifting operation that moves the bits of a binary number to the left. Unlike rotation, the bits that fall off at the end are discarded and the vacant positions are filled with zeros.

The operation:

> - Performs a left shift operation on a Field element.

Reviewer: ```suggestion
- Performs a right shift operation on a Field element.
```
